scipy/stats/_hypotests.py

```diff
@@ -1033,30 +1049,23 @@ def _pval_cvm_2samp_exact(s, nx, ny):
     """
     Compute the exact p-value of the Cramer-von Mises two-sample test
     for a given value s (float) of the test statistic by enumerating
-    all possible combinations. nx and ny are the sizes of the samples.    
+    all possible combinations. nx and ny are the sizes of the samples.
     """
-    z = np.arange(1, nx+ny+1)
-    rangex = np.arange(1, nx+1)
-    rangey = np.arange(1, ny+1)
+    rangex = np.arange(nx)
+    rangey = np.arange(ny)
 
     us = []
-    # for the first sample x, select all possible combinations of
-    # the ranks of the elements of x in the combined sample
-    # note that the acutal values of the samples are irrelevant
-    # once the ranks of the elements of x are fixed, the
-    # ranks of the elements in the second sample
-    # that z = 1, ..., nx+ny contains all possible ranks for a loop over all 
-    for c in combinations(z, nx):
-        x = np.array(c)
-        # the ranks of the second sample y are given by the set z - x
-        # note than one needs to shift by -1 since the lowest rank is 1
-        mask = np.ones(nx+ny, bool)
-        mask[x-1] = False
-        y = z[mask]
+
+    # x and y are all possible partitions of ranks from 0 to nx + ny - 1
+    # into two sets of length nx and ny
+    # Here, ranks are from 0 to nx + ny - 1 instead of 1 to nx + ny, but
+    # this does not change the value of the statistic.
+    for x, y in _all_partitions(nx, ny):
         # compute the statistic
         u = nx * np.sum((x - rangex)**2)
         u += ny * np.sum((y - rangey)**2)
         us.append(u)
+
     # compute the values of u and the frequencies
     u, cnt = np.unique(us, return_counts=True)
```

Comment:
If the goal here is to improve efficiency  couldn't we incrementally calculate? 

E.g. use a dictionary where the key is the unique value of `us` and `+=1` the value if key exists and add a new key with value `1` if the new `u` does not exist as a key.

also, I'm a bit confused here, isn't this supposed to be a change to t test but it seems like a change to cvm?

Reply:
> also, I'm a bit confused here, isn't this supposed to be a change to t test but it seems like a change to cvm?

Changes both. The code I needed in the permutation t-test was written in the CVM test. I moved it into its own function `_all_partitions`.

Reply:
>E.g. use a dictionary where the key is the unique value of us and +=1 the value if key exists and add a new key with value 1 if the new u does not exist as a key.

We're not speeding up CVM test. We're only speeding up permutation t-test. I'm just borrowing the code from CVM test because it's useful in permutation t-test - and really can be used for calculating a lot of exact distributions the naive way (using permutations).